File: leetcode/lc-us/py/2537.py
from collections import defaultdict

from typing import List


class Solution:
    def countGood(self, nums: List[int], k: int) -> int:
        res, pairs = 0, 0
        cnt = defaultdict(int)
        l, r = 0, 0
        while r < len(nums):
            pairs += cnt[nums[r]]
            cnt[nums[r]] += 1
            r += 1
            while pairs >= k:
                pairs -= cnt[nums[l]] - 1
                cnt[nums[l]] -= 1
                l += 1
            res += l
        return res

File: leetcode/lc-us/cpp/2537.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    long long countGood(vector<int> &nums, int k) {
        using LL = long long;
        LL res = 0;
        int pairs = 0;
        unordered_map<int, int> cnt;
        for (int l = 0, r = 0; r < nums.size(); r++) {
            pairs += cnt[nums[r]];
            cnt[nums[r]]++;
            while (pairs >= k) {
                pairs -= cnt[nums[l]] - 1;
                cnt[nums[l]]--;
                l += 1;
            }
            res += l;
        }
        return res;
    }
};